File: java/ch17-03-decimal/src/main/java/com/thesoftwaregorilla/tdd/money/CurrencyTransaction.java
package com.thesoftwaregorilla.tdd.money;

import java.math.BigDecimal;
import java.math.RoundingMode;

public class CurrencyTransaction {
    private final Bank bank;
    private final Money sourceAmount;
    private final String targetCurrency;

    private Money sourceFee;
    private BigDecimal targetCurrencyRateFeePercentage;
    private BigDecimal targetConversionRate;
    private Money targetAmountAfterRateFee;
    private Money targetCurrencyFee;


    private Money targetServiceFee;
    private Money totalTargetFees;
    private Money totalTransactionFees;
    private Money totalTransactionAmount;
    private Money settlementAmount;
    private boolean isSettled;

    public CurrencyTransaction(Bank bank, Money sourceAmount, String targetCurrency) {
        this.bank = bank;
        this.sourceAmount = sourceAmount;
        this.targetCurrency = targetCurrency;
        isSettled = false;
        sourceFee = new Money(BigDecimal.ZERO, sourceAmount.getCurrency());
        targetCurrencyRateFeePercentage = BigDecimal.ZERO;
        targetConversionRate = BigDecimal.ZERO;
        targetAmountAfterRateFee = new Money(BigDecimal.ZERO, targetCurrency);
        targetCurrencyFee = new Money(BigDecimal.ZERO, targetCurrency);
        targetServiceFee = new Money(BigDecimal.ZERO, targetCurrency);
        totalTargetFees = new Money(BigDecimal.ZERO, targetCurrency);
        totalTransactionFees = new Money(BigDecimal.ZERO, targetCurrency);
        totalTransactionAmount = new Money(BigDecimal.ZERO, sourceAmount.getCurrency());
        settlementAmount = new Money(BigDecimal.ZERO, targetCurrency);
    }

    public void setSourceFee(Money fee) {
        this.sourceFee = fee;
    }

    public Money getSourceFee() {
        return sourceFee;
    }

    public void setTargetCurrencyRateFeePercentage(BigDecimal rate) {
        this.targetCurrencyRateFeePercentage = rate;
    }

    public void setTargetServiceFee(Money fee) {
        this.targetServiceFee = fee;
    }

    public boolean isSettled() {
        return isSettled;
    }

    public void settle() {
        if (isSettled) {
            return;
        }
        BigDecimal bankRate = bank.rate(sourceAmount.getCurrency(),targetCurrency);
        BigDecimal rateFee = bankRate.multiply(targetCurrencyRateFeePercentage).setScale(8, RoundingMode.HALF_UP);
        this.targetConversionRate = bankRate.subtract(rateFee).setScale(8, RoundingMode.HALF_UP);
        this.targetAmountAfterRateFee = new Money(sourceAmount.getAmount().multiply(targetConversionRate), targetCurrency);
        Money srcInDest = sourceAmount.reduce(bank, targetCurrency);
        this.targetCurrencyFee = new Money(srcInDest.getAmount().subtract(targetAmountAfterRateFee.getAmount()), targetCurrency);
        this.totalTargetFees = targetCurrencyFee.plus(targetServiceFee).reduce(bank, targetCurrency);
        this.totalTransactionFees = sourceFee.plus(totalTargetFees).reduce(bank, targetCurrency);
        this.totalTransactionAmount = sourceAmount.plus(this.sourceFee).reduce(bank, sourceAmount.getCurrency());
        this.settlementAmount = new Money(this.targetAmountAfterRateFee.getAmount().subtract(targetServiceFee.getAmount()), targetCurrency);
        isSettled = true;
    }

    public BigDecimal getTargetConversionRate() {
        return targetConversionRate;
    }

    public Money getSettlementAmount() {
        return settlementAmount;
    }

    public Money getTargetCurrencyFee() {
        return targetCurrencyFee;
    }

    public Money getTargetServiceFee() {
        return targetServiceFee;
    }

    public Money getTargetAmountAfterRateFee() {
        return targetAmountAfterRateFee;
    }

    public Money getTotalTargetFees() {
        return totalTargetFees;
    }

    public Money getTotalTransactionFees() {
        return totalTransactionFees;
    }

    public Money getTotalTransactionAmount() {
        return totalTransactionAmount;
    }

    public BigDecimal getTargetCurrencyRateFeePercentage() {
        return targetCurrencyRateFeePercentage;
    }


}

File: cs/ch17-03-decimal/Money/CurrencyTransaction.cs
namespace TheSoftwareGorilla.TDD.Money;

public class CurrencyTransaction
{
    private Bank Bank { get; }
    public Money SourceAmount { get; }
    public string TargetCurrency { get; }

    public bool IsSettled { get; private set; }

    public CurrencyTransaction(Bank bank, Money sourceAmount, string targetCurrency)
    {
        Bank = bank;
        SourceAmount = sourceAmount;
        TargetCurrency = targetCurrency;
        IsSettled = false;
        SourceFee = new Money(0m, SourceAmount.Currency);
        TargetCurrencyRateFeePercentage = 0m;
        TargetServiceFee = new Money(0m, targetCurrency);
        TargetConversionRate = 0m;
        TargetCurrencyRateFeePercentage = 0m;
        TargetServiceFee = new Money(0m, targetCurrency);
        TargetConversionRate = 0m;
        TargetCurrencyFee = new Money(0m, targetCurrency);
        TotalTargetFees = new Money(0m, targetCurrency);
        SettlementAmount = new Money(0m, targetCurrency);
        TotalTransactionFees = new Money(0m, targetCurrency);
        TotalTransactionAmount = new Money(0m, SourceAmount.Currency);
        TargetAmountAfterRateFee = new Money(0m, targetCurrency);
    }

    public Money SourceFee { get; set; }
    public decimal TargetCurrencyRateFeePercentage { get; set; }
    public Money TargetServiceFee { get; set; }
    public decimal TargetConversionRate { get; private set; }
    public Money TargetCurrencyFee { get; private set; }
    public Money TotalTargetFees { get; private set; }
    public Money SettlementAmount { get; private set; }
    public Money TotalTransactionFees { get; private set; }
    public Money TotalTransactionAmount { get; private set; }
    public Money TargetAmountAfterRateFee { get; private set; }

    public void Settle()
    {
        if (IsSettled) {
            return;
        }
        decimal bankRate = Bank.Rate(SourceAmount.Currency, TargetCurrency);
        TargetConversionRate = Math.Round(bankRate - bankRate * TargetCurrencyRateFeePercentage, 8, MidpointRounding.AwayFromZero);
        TargetAmountAfterRateFee = new Money(SourceAmount.Amount * TargetConversionRate, TargetCurrency);
        Money srcInDest = SourceAmount.Reduce(Bank, TargetCurrency);
        TargetCurrencyFee = new Money(srcInDest.Amount - TargetAmountAfterRateFee.Amount, TargetCurrency);
        TotalTargetFees = TargetCurrencyFee.Plus(TargetServiceFee).Reduce(Bank, TargetCurrency);
        TotalTransactionFees = SourceFee.Plus(TotalTargetFees).Reduce(Bank, TargetCurrency);
        TotalTransactionAmount = SourceAmount.Plus(SourceFee).Reduce(Bank, SourceAmount.Currency);
        SettlementAmount = new Money(TargetAmountAfterRateFee.Amount - TargetServiceFee.Amount, TargetCurrency);
        IsSettled = true;
    }

}